File: backend/arrlink/api/vocabulary.py
"""Vocabulary CRUD/read surface: known values per rule-condition category,
and "refresh now" force-triggers for syncs that otherwise already run
automatically in the background (see core/poller.py, core/vocabulary.py).
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query

from ..core.vocabulary import RICH_CATEGORIES, sync_tmdb_vocabulary, sync_trash_vocabulary
from ..deps import get_db
from ..state import State
from .auth import CurrentUser

router = APIRouter(prefix="/api", tags=["vocabulary"])


@router.get("/vocabulary")
def get_vocabulary(
    _user: CurrentUser,
    category: str = Query(...),
    app_type: str = Query(...),
    app_id: int | None = Query(None),
    db: State = Depends(get_db),
) -> list[dict]:
    """Known values for one (category, app_type[, app_id]) scope: the
    shared vocabulary plus this app's own instance-scoped rows, if any."""
    if category not in RICH_CATEGORIES:
        raise HTTPException(422, f"category must be one of {sorted(RICH_CATEGORIES)}")
    rows = db.query(
        "SELECT value, source, external_id, app_id FROM vocabulary WHERE "
        "category=? AND app_type=? AND (app_id IS NULL OR app_id=?) ORDER BY value",
        (category, app_type, app_id),
    )
    return [dict(r) for r in rows]


@router.post("/vocabulary/import/tmdb")
def import_tmdb(_user: CurrentUser, db: State = Depends(get_db)) -> dict:
    """Force-trigger the same TMDB genre/certification sync that
    Poller._vocabulary_loop already runs automatically on a daily cadence —
    useful right after setting a TMDB key override, not required otherwise."""
    import asyncio

    counts = asyncio.run(sync_tmdb_vocabulary(db))
    if not counts:
        raise HTTPException(
            422,
            "no TMDB API key configured (set one in Settings, or the "
            "operator can configure a default via the TMDB_API_KEY env var)",
        )
    return {"imported": counts}


@router.post("/vocabulary/import/trash")
def import_trash(
    _user: CurrentUser,
    app_type: str = Query(...),
    db: State = Depends(get_db),
) -> dict:
    """Force-trigger the TRaSH Guides quality-naming sync for one app type."""
    import asyncio

    if app_type not in ("radarr", "sonarr"):
        raise HTTPException(422, "app_type must be 'radarr' or 'sonarr'")
    n = asyncio.run(sync_trash_vocabulary(db, app_type))
    return {"imported": n}


@router.post("/apps/{app_id}/vocabulary/sync")
def sync_app_vocabulary(app_id: int, _user: CurrentUser, db: State = Depends(get_db)) -> dict:
    """Force-trigger this app's per-instance vocabulary sync (quality
    profiles, languages, observed collection names) — Poller already runs
    this automatically every poll; this is just for immediate feedback."""
    import asyncio

    from ..arr.base import AdapterError
    from ..arr.factory import get_adapter
    from ..core.poller import Poller

    row = db.query_one("SELECT * FROM apps WHERE id=?", (app_id,))
    if not row:
        raise HTTPException(404, "app not found")
    try:
        adapter = get_adapter(row["type"], row["url"], row["api_key"])
        items = asyncio.run(adapter.fetch_items())
    except (ValueError, AdapterError) as e:
        detail = getattr(e, "detail", None) or str(e)
        raise HTTPException(502, detail) from e
    # Reuse Poller's instance-sync logic without needing an active Poller
    # instance/event loop — it only touches `self.db`.
    poller = Poller.__new__(Poller)
    poller.db = db
    poller._sync_instance_vocabulary(app_id, row["type"], items)
    return {"ok": True}

File: backend-ts/src/api/vocabulary.ts
import { and, eq, isNull, or } from 'drizzle-orm'
import type { FastifyInstance } from 'fastify'
import { AdapterError } from '../arr/types.js'
import { getAdapter } from '../arr/factory.js'
import type { Settings } from '../config/env.js'
import type { DbClient } from '../db/client.js'
import { apps, vocabulary as vocabularyTable } from '../db/schema.js'
import type { SettingsStore } from '../db/settings-store.js'
import {
  RICH_CATEGORIES,
  syncTmdbVocabulary,
  syncTrashVocabulary,
  syncVocabularyRows,
} from '../core/vocabulary.js'
import { HttpError } from '../http-error.js'
import { getCurrentUser } from './auth.js'

/** Vocabulary read surface + "refresh now" triggers for the poller's
 * background syncs. Ported from api/vocabulary.py. Wire format is
 * snake_case, matching web/ and the Python backend. */

export interface VocabularyRouteOptions {
  db: DbClient
  settingsStore: SettingsStore
  env: Settings
}

export function registerVocabularyRoutes(
  app: FastifyInstance,
  opts: VocabularyRouteOptions,
): void {
  const { db, settingsStore, env } = opts

  // Known values for one scope: shared vocabulary plus this app's own rows.
  app.get('/api/vocabulary', (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const q = request.query as {
      category?: string
      app_type?: string
      app_id?: string
    }
    const category = q.category ?? ''
    const appType = q.app_type ?? ''
    if (!RICH_CATEGORIES.has(category)) {
      throw new HttpError(
        422,
        `category must be one of ${[...RICH_CATEGORIES].sort().join(', ')}`,
      )
    }
    const appId = q.app_id !== undefined ? Number(q.app_id) : null
    // No app_id means the rule editor is asking for a type-scoped rule (no
    // single instance) -- union across every instance of this app_type's own
    // vocabulary instead of just the shared rows, so a category like
    // "language" (which has no shared/global source at all) still returns
    // real suggestions instead of an empty dropdown.
    const scope =
      appId !== null
        ? or(isNull(vocabularyTable.appId), eq(vocabularyTable.appId, appId))
        : null
    const rows = db
      .select({
        value: vocabularyTable.value,
        source: vocabularyTable.source,
        externalId: vocabularyTable.externalId,
        appId: vocabularyTable.appId,
      })
      .from(vocabularyTable)
      .where(
        and(
          eq(vocabularyTable.category, category),
          eq(vocabularyTable.appType, appType),
          scope ?? undefined,
        ),
      )
      .orderBy(vocabularyTable.value)
      .all()
    // De-dupe by value when unioning across instances -- e.g. two Radarr
    // instances both reporting "English" shouldn't show up twice.
    const seen = new Set<string>()
    const out: Array<{
      value: string
      source: string
      external_id: string | null
      app_id: number | null
    }> = []
    for (const r of rows) {
      if (seen.has(r.value)) continue
      seen.add(r.value)
      out.push({
        value: r.value,
        source: r.source,
        external_id: r.externalId,
        app_id: r.appId,
      })
    }
    return out
  })

  // Force-trigger the poller's daily TMDB sync -- useful right after
  // setting a key override.
  app.post('/api/vocabulary/import/tmdb', async (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const counts = await syncTmdbVocabulary(db, settingsStore)
    if (Object.keys(counts).length === 0) {
      throw new HttpError(
        422,
        'no TMDB API key configured (set one in Settings, or the operator can ' +
          'configure a default via the TMDB_API_KEY env var)',
      )
    }
    return { imported: counts }
  })

  // Force-trigger the TRaSH Guides quality-naming sync for one app type.
  app.post('/api/vocabulary/import/trash', async (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const appType = (request.query as { app_type?: string }).app_type ?? ''
    if (appType !== 'radarr' && appType !== 'sonarr') {
      throw new HttpError(422, "app_type must be 'radarr' or 'sonarr'")
    }
    const n = await syncTrashVocabulary(db, appType)
    return { imported: n }
  })

  // Force-trigger this app's per-instance sync -- the poller already runs
  // it every poll; this is just for immediate feedback.
  app.post('/api/apps/:appId/vocabulary/sync', async (request) => {
    getCurrentUser(request, db, settingsStore, env)
    const appId = Number((request.params as { appId: string }).appId)
    const row = db.select().from(apps).where(eq(apps.id, appId)).get()
    if (!row) throw new HttpError(404, 'app not found')
    const adapter = getAdapter(row.type, row.url, row.apiKey)
    let items: Awaited<ReturnType<typeof adapter.fetchItems>>
    try {
      items = await adapter.fetchItems()
    } catch (e) {
      const detail = e instanceof AdapterError ? e.detail : String(e)
      throw new HttpError(502, detail)
    }
    const collections = [
      ...new Set(items.map((i) => i.collection).filter((c): c is string => Boolean(c))),
    ].sort()
    syncVocabularyRows(
      db,
      'collection',
      row.type,
      appId,
      collections.map((c): [string, null] => [c, null]),
      'observed',
    )
    const profiles = await adapter.fetchQualityProfiles()
    const languages = await adapter.fetchLanguages()
    syncVocabularyRows(
      db,
      'quality',
      row.type,
      appId,
      profiles.filter((p) => p.name).map((p): [string, string] => [p.name, String(p.id)]),
      'instance',
    )
    syncVocabularyRows(
      db,
      'language',
      row.type,
      appId,
      languages
        .filter((l) => l.name)
        .map((l): [string, string] => [l.name, String(l.id)]),
      'instance',
    )
    return { ok: true }
  })
}
